File: risk.py
import config


def position_size(capital: float, price: float, risk_pct: float = config.RISK_PER_TRADE) -> float:
    risk_amount = capital * risk_pct
    qty = risk_amount / price
    return round(qty, 6)


def stop_loss_price(entry: float, pct: float = config.STOP_LOSS_PCT) -> float:
    return round(entry * (1 - pct), 4)


def take_profit_price(entry: float, pct: float = config.TAKE_PROFIT_PCT) -> float:
    return round(entry * (1 + pct), 4)


def is_max_positions_reached(open_positions: dict) -> bool:
    return len(open_positions) >= config.MAX_OPEN_POSITIONS


def calc_pnl(entry: float, current: float, qty: float) -> float:
    return round((current - entry) * qty, 4)


def calc_pnl_pct(entry: float, current: float) -> float:
    return round((current - entry) / entry * 100, 2)

File: web/src/lib/risk.js
import { RISK_PER_TRADE, STOP_LOSS_PCT, TAKE_PROFIT_PCT, MAX_OPEN_POSITIONS } from "./config.js";

export function positionSize(capital, price, riskPct = RISK_PER_TRADE) {
  const riskAmount = capital * riskPct;
  return Math.round((riskAmount / price) * 1e6) / 1e6;
}

export function stopLossPrice(entry, pct = STOP_LOSS_PCT) {
  return Math.round(entry * (1 - pct) * 1e4) / 1e4;
}

export function takeProfitPrice(entry, pct = TAKE_PROFIT_PCT) {
  return Math.round(entry * (1 + pct) * 1e4) / 1e4;
}

export function isMaxPositionsReached(openPositions) {
  return Object.keys(openPositions).length >= MAX_OPEN_POSITIONS;
}

export function calcPnl(entry, current, qty) {
  return Math.round((current - entry) * qty * 1e4) / 1e4;
}

export function calcPnlPct(entry, current) {
  return Math.round(((current - entry) / entry) * 100 * 100) / 100;
}
